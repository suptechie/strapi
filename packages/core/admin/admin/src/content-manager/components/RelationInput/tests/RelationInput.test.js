import React from 'react';
import { IntlProvider } from 'react-intl';
import { MemoryRouter } from 'react-router-dom';
import { render, screen, fireEvent } from '@testing-library/react';
import { ThemeProvider, lightTheme } from '@strapi/design-system';

import { RelationInput } from '../index';

const FIXTURES_RELATIONS = {
  data: {
    pages: [
      [
        {
          id: 1,
          href: '/',
          mainField: 'Relation 1',
          publicationState: 'draft',
        },
        {
          id: 2,
          href: '',
          mainField: 'Relation 2',
          publicationState: 'published',
        },
        {
          id: 3,
          href: '',
          mainField: 'Relation 3',
          publicationState: false,
        },
      ],
    ],
  },
  isLoading: false,
  isSuccess: true,
  hasNextPage: true,
};

const FIXTURES_SEARCH = {
  data: {
    pages: [
      [
        {
          id: 4,
          mainField: 'Relation 4',
          publicationState: 'draft',
        },
      ],
    ],
  },
  isLoading: false,
  isSuccess: true,
};

const setup = (props) =>
  render(
    <MemoryRouter>
      <ThemeProvider theme={lightTheme}>
        <IntlProvider locale="en">
          <RelationInput
            description="this is a description"
            id="1"
            name="some-relation-1"
            label="Some Relation"
            labelLoadMore="Load more"
            loadingMessage={() => 'Relations are loading'}
            numberOfRelationsToDisplay={5}
            onRelationAdd={() => jest.fn()}
            onSearchOpen={() => jest.fn()}
            onSearchClose={() => jest.fn()}
            onRelationRemove={() => jest.fn()}
            onRelationLoadMore={() => jest.fn()}
            onSearch={() => jest.fn()}
            onSearchNextPage={() => jest.fn()}
            placeholder="Select..."
            publicationStateTranslations={{
              draft: 'Draft',
              published: 'Published',
            }}
            relations={FIXTURES_RELATIONS}
            searchResults={FIXTURES_SEARCH}
            size={8}
            {...props}
          />
        </IntlProvider>
      </ThemeProvider>
    </MemoryRouter>
  );

describe('Content-Manager || RelationInput', () => {
  test('should render and match snapshot', () => {
    const { container } = setup();

    expect(container).toMatchSnapshot();
    expect(screen.getByText('Some Relation')).toBeInTheDocument();
    expect(screen.getByText('Load more')).toBeInTheDocument();
    expect(screen.getByText('Select...')).toBeInTheDocument();
    expect(screen.getByText('Relation 1')).toBeInTheDocument();
  });

  describe('Callbacks', () => {
    test('should call onSearchOpen', () => {
      const spy = jest.fn();
      setup({ onSearchOpen: spy });

      fireEvent.mouseDown(screen.getByText(/select\.\.\./i));

      expect(spy).toHaveBeenCalled();
    });

    test('should call onRelationAdd and onSearchClose', () => {
      const onAddSpy = jest.fn();
      const onCloseSpy = jest.fn();
      setup({ onRelationAdd: onAddSpy, onSearchClose: onCloseSpy });

      fireEvent.mouseDown(screen.getByText(/select\.\.\./i));
      expect(screen.getByText('Relation 4')).toBeInTheDocument();

      fireEvent.click(screen.getByText('Relation 4'));

      expect(onAddSpy).toHaveBeenCalled();
      expect(onCloseSpy).toHaveBeenCalled();
    });

    test('should call onRelationRemove', () => {
      const spy = jest.fn();
      setup({ onRelationRemove: spy });

      fireEvent.click(screen.getByTestId('remove-relation-1'));

      expect(spy).toHaveBeenCalled();
    });

    test('should call onRelationLoadMore', () => {
      const spy = jest.fn();
      setup({ onRelationLoadMore: spy });

      fireEvent.click(screen.getByText('Load more'));

      expect(spy).toHaveBeenCalled();
    });

    test('should call onSearch', () => {
      const spy = jest.fn();
      const { container } = setup({ onSearch: spy });

      fireEvent.change(container.querySelector('input'), {
        target: { value: 'searching' },
      });

      expect(spy).toHaveBeenCalled();
    });

    // TODO: check if it is possible to fire scroll event here
    // test.only('should call onSearchNextPage', () => {
    //   const spy = jest.fn();
    //   const { container } = setup({ onSearchNextPage: spy });

    //   fireEvent.mouseDown(screen.getByText(/select\.\.\./i));
    //   fireEvent.scroll(screen.getByText('Relation 4'), {
    //     not working with scrollY either
    //     target: { scrollBottom: 100 },
    //   });

    //   expect(spy).toHaveBeenCalled();
    // });
  });

  describe('States', () => {
    test('should display search loading state', () => {
      setup({ searchResults: { data: { pages: [] }, isLoading: true, isSuccess: true } });

      fireEvent.mouseDown(screen.getByText(/select\.\.\./i));

      expect(screen.getByText('Relations are loading')).toBeInTheDocument();
    });

    test('should display load more button loading if loading is true', () => {
      setup({
        relations: { data: { pages: [] }, isLoading: true, isSuccess: true, hasNextPage: true },
      });

      expect(screen.getByRole('button', { name: /load more/i })).toHaveAttribute(
        'aria-disabled',
        'true'
      );
    });

    test('should not display load more button loading if there is no next page', () => {
      setup({
        relations: { data: { pages: [] }, isLoading: false, isSuccess: true, hasNextPage: false },
      });

      expect(screen.queryByText('Load more')).not.toBeInTheDocument();
    });

    test('should display load more button loading if there is no next page but loading is true', () => {
      setup({
        relations: { data: { pages: [] }, isLoading: true, isSuccess: true, hasNextPage: false },
      });

      expect(screen.getByRole('button', { name: /load more/i })).toHaveAttribute(
        'aria-disabled',
        'true'
      );
    });

    test('should display error state', () => {
      setup({ error: 'This is an error' });

      expect(screen.getByText('This is an error')).toBeInTheDocument();
    });

    test('should apply disabled state', () => {
      setup({ disabled: true });

      expect(screen.queryByText('Load more')).not.toBeInTheDocument();
      expect(screen.getByRole('textbox')).toBeDisabled();
      expect(screen.getByTestId('remove-relation-1')).toBeDisabled();
    });
  });
});
