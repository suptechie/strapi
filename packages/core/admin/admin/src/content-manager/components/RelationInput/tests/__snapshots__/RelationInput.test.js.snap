// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content-Manager || RelationInput should render and match snapshot 1`] = `
.c35 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c0 {
  background: #f6f6f9;
  padding-top: 16px;
  padding-right: 24px;
  padding-bottom: 16px;
  padding-left: 24px;
  border-radius: 4px;
}

.c8 {
  padding-right: 12px;
}

.c20 {
  padding-top: 4px;
  padding-right: 16px;
  padding-bottom: 4px;
  min-width: 0px;
}

.c27 {
  padding-left: 16px;
}

.c33 {
  padding-top: 8px;
}

.c7 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c6 {
  font-weight: 600;
  color: #32324d;
  font-size: 0.75rem;
  line-height: 1.33;
}

.c34 {
  color: #666687;
  font-size: 0.75rem;
  line-height: 1.33;
}

.c1 {
  padding-bottom: 12px;
}

.c17 {
  background: #ffffff;
  padding-top: 12px;
  padding-right: 16px;
  padding-bottom: 12px;
  padding-left: 16px;
  border-radius: 4px;
  border-color: #dcdce4;
  border: 1px solid #dcdce4;
}

.c2 {
  -webkit-align-items: end;
  -webkit-box-align: end;
  -ms-flex-align: end;
  align-items: end;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  gap: 8px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c11 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c18 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c28 {
  color: #666687;
  width: 12px;
}

.c29 path {
  fill: #666687;
}

.c3 {
  -webkit-flex-basis: 70%;
  -ms-flex-preferred-size: 70%;
  flex-basis: 70%;
}

.c4 {
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c5 > * {
  margin-top: 0;
  margin-bottom: 0;
}

.c5 > * + * {
  margin-top: 4px;
}

.c30 {
  color: #4945ff;
  display: block;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  font-size: 0.875rem;
  line-height: 1.43;
}

.c9 {
  background: transparent;
  border: none;
  position: relative;
  z-index: 1;
}

.c9 svg {
  height: 0.6875rem;
  width: 0.6875rem;
}

.c9 svg path {
  fill: #666687;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: none;
  border: none;
}

.c10 svg {
  width: 0.5625rem;
}

.c24 {
  background: #eaf5ff;
  padding: 4px;
  border-radius: 4px;
  border-color: #b8e1ff;
  border: 1px solid #b8e1ff;
  min-width: 20px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c31 {
  background: #eafbe7;
  padding: 4px;
  border-radius: 4px;
  border-color: #c6f0c2;
  border: 1px solid #c6f0c2;
  min-width: 20px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c25 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c26 {
  color: #006096;
  font-weight: 600;
  font-size: 0.6875rem;
  line-height: 1.45;
  text-transform: uppercase;
}

.c32 {
  color: #2f6846;
  font-weight: 600;
  font-size: 0.6875rem;
  line-height: 1.45;
  text-transform: uppercase;
}

.c23 {
  color: #4945ff;
  font-size: 0.875rem;
  line-height: 1.43;
}

.c21 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  position: relative;
  outline: none;
}

.c21 svg path {
  fill: #4945ff;
}

.c21 svg {
  font-size: 0.625rem;
}

.c21:after {
  -webkit-transition-property: all;
  transition-property: all;
  -webkit-transition-duration: 0.2s;
  transition-duration: 0.2s;
  border-radius: 8px;
  content: '';
  position: absolute;
  top: -4px;
  bottom: -4px;
  left: -4px;
  right: -4px;
  border: 2px solid transparent;
}

.c21:focus-visible {
  outline: none;
}

.c21:focus-visible:after {
  border-radius: 8px;
  content: '';
  position: absolute;
  top: -5px;
  bottom: -5px;
  left: -5px;
  right: -5px;
  border: 2px solid #4945ff;
}

.c14 {
  padding-right: 8px;
}

.c15 {
  color: #4945ff;
  font-size: 0.75rem;
  line-height: 1.33;
}

.c12 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c13 {
  background: transparent;
  border: none;
  position: relative;
  outline: none;
}

.c13[aria-disabled='true'] {
  pointer-events: none;
}

.c13[aria-disabled='true'] svg path {
  fill: #666687;
}

.c13 svg {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 0.625rem;
}

.c13 svg path {
  fill: #4945ff;
}

.c13:after {
  -webkit-transition-property: all;
  transition-property: all;
  -webkit-transition-duration: 0.2s;
  transition-duration: 0.2s;
  border-radius: 8px;
  content: '';
  position: absolute;
  top: -4px;
  bottom: -4px;
  left: -4px;
  right: -4px;
  border: 2px solid transparent;
}

.c13:focus-visible {
  outline: none;
}

.c13:focus-visible:after {
  border-radius: 8px;
  content: '';
  position: absolute;
  top: -5px;
  bottom: -5px;
  left: -5px;
  right: -5px;
  border: 2px solid #4945ff;
}

.c19 {
  width: 100%;
  min-width: 0;
}

.c16 {
  position: relative;
  overflow-x: hidden;
  overflow-y: auto;
}

.c16:before,
.c16:after {
  position: absolute;
  width: 100%;
  height: 4px;
  z-index: 1;
}

.c16:before {
  background: linear-gradient(rgba(33,33,52,0.1) 0%,rgba(0,0,0,0) 100%);
  top: 0;
}

.c16:after {
  background: linear-gradient(rgba(0,0,0,0) 0%,rgba(33,33,52,0.1) 100%);
  bottom: 0;
}

.c22 {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  display: inherit;
}

<div>
  <div>
    <div
      class="c0"
    >
      <div
        class="c1 c2"
        wrap="wrap"
      >
        <div
          class="c3 c4 c5"
          spacing="1"
        >
          <label
            class="c6"
            for="1"
          >
            <div
              class="c7"
            >
              Some Relation
            </div>
          </label>
          <div
            class=" css-1tkrshn-container"
          >
            <div
              class=" css-1sb7bue-control"
            >
              <div
                class=" css-1kj7tu4-ValueContainer"
              >
                <div
                  class=" css-6gbuim-placeholder"
                >
                  Select...
                </div>
                <div
                  class="css-fph3ax-Input"
                >
                  <div
                    class=""
                    style="display: inline-block;"
                  >
                    <input
                      aria-autocomplete="list"
                      autocapitalize="none"
                      autocomplete="off"
                      autocorrect="off"
                      id="1"
                      spellcheck="false"
                      style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                      tabindex="0"
                      type="text"
                      value=""
                    />
                    <div
                      style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                    />
                  </div>
                </div>
              </div>
              <div
                class=" css-1hb7zxy-IndicatorsContainer"
              >
                <button
                  class="c8 c9 c10"
                  type="button"
                >
                  <svg
                    fill="none"
                    height="1em"
                    viewBox="0 0 14 8"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M14 .889a.86.86 0 01-.26.625L7.615 7.736A.834.834 0 017 8a.834.834 0 01-.615-.264L.26 1.514A.861.861 0 010 .889c0-.24.087-.45.26-.625A.834.834 0 01.875 0h12.25c.237 0 .442.088.615.264a.86.86 0 01.26.625z"
                      fill="#32324D"
                      fill-rule="evenodd"
                    />
                  </svg>
                </button>
              </div>
            </div>
            <input
              name="some-relation-1"
              type="hidden"
              value=""
            />
          </div>
        </div>
        <div
          class="c11"
        >
          <button
            aria-disabled="false"
            class="c12 c13"
            type="button"
          >
            <span
              aria-hidden="true"
              class="c14"
            >
              <svg
                fill="none"
                height="1em"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M15.681 2.804A9.64 9.64 0 0011.818 2C6.398 2 2 6.48 2 12c0 5.521 4.397 10 9.818 10 2.03 0 4.011-.641 5.67-1.835a9.987 9.987 0 003.589-4.831 1.117 1.117 0 00-.664-1.418 1.086 1.086 0 00-1.393.676 7.769 7.769 0 01-2.792 3.758 7.546 7.546 0 01-4.41 1.428V4.222h.002a7.492 7.492 0 013.003.625 7.61 7.61 0 012.5 1.762l.464.551-2.986 3.042a.186.186 0 00.129.316H22V3.317a.188.188 0 00-.112-.172.179.179 0 00-.199.04l-2.355 2.4-.394-.468-.02-.02a9.791 9.791 0 00-3.239-2.293zm-3.863 1.418V2v2.222zm0 0v15.556c-4.216 0-7.636-3.484-7.636-7.778s3.42-7.777 7.636-7.778z"
                  fill="#212134"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
            <span
              class="c15"
            >
              Load more
            </span>
          </button>
        </div>
      </div>
      <div
        class=""
      >
        <div
          class="c16"
        >
          <div
            class="c4 c5"
            spacing="1"
          >
            <div
              style="position: relative; height: 174px; overflow: auto; will-change: transform; direction: ltr;"
            >
              <ol
                style="height: 174px; width: 100%;"
              >
                <li
                  class=""
                  style="position: absolute; left: 0px; top: 0px; height: 58px; width: 100%;"
                >
                  <div
                    class="c17 c18"
                  >
                    <div
                      class="c18 c19"
                    >
                      <div
                        class="c20"
                      >
                        <a
                          aria-current="page"
                          class="c21 c22 active"
                          href="/"
                        >
                          <span
                            class="c23"
                          >
                            Relation 1
                          </span>
                        </a>
                      </div>
                      <div
                        class="c24 c25"
                      >
                        <span
                          class="c26"
                        >
                          Draft
                        </span>
                      </div>
                    </div>
                    <div
                      class="c27"
                    >
                      <button
                        data-testid="remove-relation-1"
                        type="button"
                      >
                        <svg
                          class="c28 c29"
                          fill="none"
                          height="1em"
                          viewBox="0 0 24 24"
                          width="12px"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M24 2.417L21.583 0 12 9.583 2.417 0 0 2.417 9.583 12 0 21.583 2.417 24 12 14.417 21.583 24 24 21.583 14.417 12 24 2.417z"
                            fill="#212134"
                          />
                        </svg>
                      </button>
                    </div>
                  </div>
                </li>
                <li
                  class=""
                  style="position: absolute; left: 0px; top: 58px; height: 58px; width: 100%;"
                >
                  <div
                    class="c17 c18"
                  >
                    <div
                      class="c18 c19"
                    >
                      <div
                        class="c20"
                      >
                        <span
                          class="c30"
                        >
                          Relation 2
                        </span>
                      </div>
                      <div
                        class="c31 c25"
                      >
                        <span
                          class="c32"
                        >
                          Published
                        </span>
                      </div>
                    </div>
                    <div
                      class="c27"
                    >
                      <button
                        data-testid="remove-relation-2"
                        type="button"
                      >
                        <svg
                          class="c28 c29"
                          fill="none"
                          height="1em"
                          viewBox="0 0 24 24"
                          width="12px"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M24 2.417L21.583 0 12 9.583 2.417 0 0 2.417 9.583 12 0 21.583 2.417 24 12 14.417 21.583 24 24 21.583 14.417 12 24 2.417z"
                            fill="#212134"
                          />
                        </svg>
                      </button>
                    </div>
                  </div>
                </li>
                <li
                  class=""
                  style="position: absolute; left: 0px; top: 116px; height: 58px; width: 100%;"
                >
                  <div
                    class="c17 c18"
                  >
                    <div
                      class="c18 c19"
                    >
                      <div
                        class="c20"
                      >
                        <span
                          class="c30"
                        >
                          Relation 3
                        </span>
                      </div>
                    </div>
                    <div
                      class="c27"
                    >
                      <button
                        data-testid="remove-relation-3"
                        type="button"
                      >
                        <svg
                          class="c28 c29"
                          fill="none"
                          height="1em"
                          viewBox="0 0 24 24"
                          width="12px"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M24 2.417L21.583 0 12 9.583 2.417 0 0 2.417 9.583 12 0 21.583 2.417 24 12 14.417 21.583 24 24 21.583 14.417 12 24 2.417z"
                            fill="#212134"
                          />
                        </svg>
                      </button>
                    </div>
                  </div>
                </li>
              </ol>
            </div>
          </div>
        </div>
        <div
          class="c33"
        >
          <p
            class="c34"
            id="1-hint"
          >
            this is a description
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    class="c35"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </div>
</div>
`;
